src/scancode/resource.py

```diff
@@ -1497,26 +1497,75 @@ def get_resource_basic_attributes(_resource_data):
+                if field in base_fields:
+                    # We only want the fields that are not part of the base set of fields
+                    continue
+                value = field.default
+                if isinstance(value, attr.Factory):
+                    # For fields that have Factories as values, we set their values to be an
+                    # instance of whatever type the factory makes
+                    value = value.factory()
+                resource_data[field.name] = value
+            return resource_data
+
+        # Create root resource without setting root data just yet. If we run into the root data
+        # while we iterate through `resources_data`, we fill in the data then.
+        sample_resource_path = sample_resource_data.get('path')
+        root_path = sample_resource_path.split('/')[0]
+        root_name = root_path
+        root_is_file = False
+        root_data = create_empty_resource_data()
         root_resource = self._create_root_resource(root_name, root_path, root_is_file, root_data)
 
         # To help recreate the resource tree we keep a mapping by path of any
         # parent resource
         parent_by_path = {root_path: root_resource}
 
-        # resources data MUST be in top-down order
+        def get_or_create_parent(path, parent_by_path):
+            """
+            Return a parent resource for a given `path` from `parent_by_path`.
+
+            If a parent resource for a `path` does not exist in `parent_by_path`, it is created recursively.
+
+            Note: the root path and root Resource must already be in `parent_by_path` or else this
+            function does not work.
+            """
+            parent_path = parent_directory(path).rstrip('/')
+            if parent_path in parent_by_path:
+                return parent_by_path[parent_path]
+            parent_parent = get_or_create_parent(parent_path, parent_by_path)
+            parent_name = file_base_name(parent_path)
+            parent_is_file = False
+            parent_resource_data = create_empty_resource_data()
+            parent_resource = self._create_resource(parent_name, parent_parent, parent_is_file, parent_path, parent_resource_data)
+            parent_by_path[parent_path] = parent_resource
+            return parent_resource
+
         for resource_data in resources_data:
             name, path, is_file = get_resource_basic_attributes(resource_data)
-            parent_path = parent_directory(path).rstrip('/')
-            # this must succeed since we must be in the proper sort order
-            parent = parent_by_path[parent_path]
-            resource = self._create_resource(name, parent, is_file, path, resource_data)
-
-            # Files are not parents (for now), so we do not need to add this
-            # to the parent_by_path mapping
-            if not is_file:
-                parent_by_path[path] = resource
+            if path in parent_by_path:
```

Comment:
Same comment as above, parent_by_path.get would be better to avoid extra lookup without loosing any readability IMHO